File: src/Avalonia.Samples/CompleteApps/Avalonia.MusicStore/ViewModels/AlbumViewModel.cs
using System;
using System.Threading.Tasks;
using Avalonia.Media.Imaging;
using Avalonia.MusicStore.Models;
using Avalonia.MusicStore.Services;

namespace Avalonia.MusicStore.ViewModels
{
    public partial class AlbumViewModel : ViewModelBase, IEquatable<AlbumViewModel>
    {
        private static readonly AlbumService s_albumService = new();
        private readonly Album _album;

        public AlbumViewModel(Album album)
        {
            _album = album;
        }

        public string Artist => _album.Artist;

        public string Title => _album.Title;

        public Task<Bitmap?> Cover => LoadCoverAsync();

        /// <summary>
        /// Asynchronously loads and decodes the album cover image, then assigns it to <see cref="Cover"/>.
        /// </summary>
        private async Task<Bitmap?> LoadCoverAsync()
        {
            try
            {
                // We wait a few ms to demonstrate that the images are loaded in the background. 
                // Remove this line in production.
                await Task.Delay(200);
                
                await using (var imageStream = await s_albumService.LoadCoverBitmapAsync(_album))
                {
                    return await Task.Run(() => Bitmap.DecodeToWidth(imageStream, 400));
                }
            }
            catch
            {
                return null;
            }
        }

        /// <summary>
        /// Saves the album and its cover to cache.
        /// </summary>        
        public async Task SaveToDiskAsync()
        {
            await s_albumService.SaveAsync(_album);

            if (await LoadCoverAsync() is { } cover)
            {
                await Task.Run(() =>
                {
                    using (var fs = s_albumService.SaveCoverBitmapStream(_album))
                    {
                        cover.Save(fs);
                    }
                });
            }
        }

        public bool Equals(AlbumViewModel? other)
        {
            if (other is null) return false;
            if (ReferenceEquals(this, other)) return true;
            return _album.Equals(other._album);
        }

        public override bool Equals(object? obj)
        {
            if (obj is null) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != GetType()) return false;
            return Equals((AlbumViewModel)obj);
        }

        public override int GetHashCode()
        {
            return _album.GetHashCode();
        }
    }
}

File: src/Avalonia.Samples/FSharp/CompleteApps/MusicStore/ViewModels/AlbumViewModel.fs
namespace MusicStore.ViewModels

open System.Diagnostics
open Avalonia.Media.Imaging
open Avalonia.Threading
open System
open System.Threading.Tasks
open MusicStore.Models

type AlbumViewModel(album: Album) =
    inherit ViewModelBase()
    
    let _album = album
    let mutable _cover: Bitmap | null = null
    
    member _.Artist
        with get() = _album.Artist

    member _.Title
        with get() = _album.Title
        
    member _.Cover
        with get() = _cover
        and private set value =
            if value <> _cover then
                // dispose previous bitmap to free native resources
                match _cover with
                | null -> ()
                | oldBmp -> (oldBmp :> IDisposable).Dispose()
                _cover <- value
                base.OnPropertyChanged("Cover")
        
    member this.LoadCover(?width: int) =
        task {
            try
                let width = defaultArg width 400
                use! imageStream = _album.LoadCoverBitmapAsync()
                let! bmp = Task.Run(fun () -> Bitmap.DecodeToWidth(imageStream, width))
                // Ensure property change is raised on the UI thread so the Image.Source updates correctly
                Dispatcher.UIThread.Post(fun () -> this.Cover <- bmp)
            with exn ->
                Debug.WriteLine $"Failed to load album cover: %s{exn.Message}"
        }
        
    member this.SaveToDiskAsync() =
        task {
            do! _album.SaveAsync();
    
            if not (isNull this.Cover) then
                let bitmap = this.Cover
                do! Task.Run(fun () ->
                    use fs = _album.SaveCoverBitmapStream()
                    bitmap.Save(fs)
                )
        }
